rust/rope/src/find.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of string finding in ropes.
+
+use rope::RopeInfo;
+use tree::Cursor;
+
+use memchr::{memchr, memchr2, memchr3};
+
+pub enum FindResult {
+    Found(usize),
+    NotFound,
+    TryAgain,
+}
+
+/// Case-sensitive find (just does byte comparison, no Unicode logic).
+///
+/// Can panic if `pat` is empty.
+pub fn find(cursor: &mut Cursor<RopeInfo>, pat: &str) -> Option<usize> {
```

Comment:
I think your understanding is basically correct, except I'm not sure what a public `findall` would do; as it is, repeated calls to `find` behave a lot like an iterator, which seems like a nice way to expose that functionality.

I _could_ add an `upto_offset` argument, but there are two other ways to achieve the same functionality. First, you could substring the rope before creating the cursor (this does require a little subtlety about not slicing a codepoint in half, but is doable and efficient). Second, you could call `find_progress` and then discard the result if it's past your range.